discovery_engine_core/core/src/stack/exploration.rs

```diff
@@ -0,0 +1,531 @@
+        negative_cois,
+        documents,
+        config,
+        &mut rand::thread_rng(),
+    )
+}
+
+fn document_selection_with_rng<R>(
+    positive_cois: &[PositiveCoi],
+    negative_cois: &[NegativeCoi],
+    documents: Vec<Document>,
+    config: &Config,
+    rng: &mut R,
+) -> Result<Vec<Document>, Error>
+where
+    R: Rng + ?Sized,
+{
+    if positive_cois.len() + negative_cois.len() == 0 {
+        return Err(Error::NotEnoughCois);
+    }
+
+    let document_embeddings = documents
+        .iter()
+        .map(|document| document.smbert_embedding.view())
+        .collect_vec();
+
+    let pos_cois = positive_cois.iter().map(|coi| coi.point().view());
+    let neg_cois = negative_cois.iter().map(|coi| coi.point().view());
+
+    let cois = chain!(pos_cois, neg_cois).collect_vec();
+    // find_nearest_coi_for_docs can't panic because we make sure beforehand
+    // that both positive and negative cois aren't empty
+    let nearest_coi_for_docs = find_nearest_coi_for_docs(&document_embeddings, &cois);
+    let doc_similarities = pairwise_cosine_similarity(document_embeddings.into_iter());
+
+    let selected = select_by_randomization_with_threshold(
+        &doc_similarities,
+        &nearest_coi_for_docs,
+        config.number_of_candidates,
+        config.max_selected_docs,
+        rng,
+    );
+
+    Ok(retain_documents_by_indices(selected, documents))
+}
+
+fn select_by_randomization_with_threshold<R>(
+    doc_similarities: &Array2<f32>,
+    nearest_coi_for_docs: &[f32],
+    number_of_candidates: usize,
+    max_selected_docs: usize,
+    rng: &mut R,
+) -> Vec<usize>
+where
+    R: Rng + ?Sized,
+{
+    let (threshold, mut candidates) =
+        select_initial_candidates(nearest_coi_for_docs, number_of_candidates);
+
+    let mut selected = Vec::new();
+    while !candidates.is_empty() && selected.len() < max_selected_docs {
+        let chosen_doc = *candidates
+            .iter()
+            .choose(rng)
+            .unwrap(/* the condition of the `while` ensures that `candidates` can't be empty */);
+
+        // remove all docs that have a similarity to chosen_doc >= threshold
+        let to_remove = doc_similarities
+            .row(chosen_doc)
+            .iter()
+            .enumerate()
+            .filter(|(_, doc_sim)| **doc_sim >= threshold)
+            .map(|(idx, _)| idx)
+            .collect();
+        candidates = &candidates - &to_remove;
+
+        selected.push(chosen_doc);
+    }
+
+    selected
+}
+
+/// Returns the indices that would sort an array.
+fn argsort(arr: &[f32]) -> Vec<usize> {
+    let mut indices: Vec<usize> = (0..arr.len()).collect();
+    indices.sort_unstable_by(move |&i, &j| nan_safe_f32_cmp(&arr[i], &arr[j]));
+    indices
+}
+
+/// Computes the cosine similarity between the cois and documents and returns the
+/// cosine similarity of the nearest coi for each document.
+///
+/// # Panics
+/// Panics if `cois` is empty.
+fn find_nearest_coi_for_docs(
```

Comment:
:art:  the name is slightly misleading as we don't return CoIs but similarities.